# Assignment 3 - Clustering town councils

The **objective of the assignment** is to find the best way to cluster the town councils, based on the budgets.


## Data Reading and Preparation
Before investigating the data, we need to read the file and get only the rows we need.

We first read the data, and then we keep only the Town Council data:

In [75]:
import pandas as pd
import numpy as np

def get_column_match_string(data, column, string):
    return data[data[column].str.contains(string)]

town_council_data = pd.read_csv("budgets/Pressupostos_dels_ens_municipals_de_Catalunya.csv")
town_council_data = get_column_match_string(town_council_data, "NOM_COMPLERT", "Ajuntament")

town_council_data

ANY_EXERCICI TIPUS_PARTIDA TIPUS_CLASSIF  CODI_PANTALLA  \
245     01/01/2016 12:00:00 AM             D             F          33.00   
671     01/01/2016 12:00:00 AM             D             F          34.00   
1501    01/01/2016 12:00:00 AM             D             F         425.00   
2325    01/01/2016 12:00:00 AM             I             E           1.00   
2326    01/01/2016 12:00:00 AM             I             E          11.00   
2327    01/01/2016 12:00:00 AM             I             E         112.00   
2328    01/01/2016 12:00:00 AM             I             E         113.00   
2329    01/01/2016 12:00:00 AM             I             E         115.00   
2330    01/01/2016 12:00:00 AM             I             E         116.00   
2331    01/01/2016 12:00:00 AM             I             E          13.00   
2332    01/01/2016 12:00:00 AM             I             E         130.00   
2333    01/01/2016 12:00:00 AM             I             E           2.00   
2334    01/01/2016 12:00:00 AM             I             E          29.00   
2335    01/01/2016 12:00:00 AM             I             E         290.00   
2336    01/01/2016 12:00:00 AM             I             E           3.00   
2337    01/01/2016 12:00:00 AM             I             E          30.00   
2338    01/01/2016 12:00:00 AM             I             E         300.00   
2339    01/01/2016 12:00:00 AM             I             E         304.00   
2340    01/01/2016 12:00:00 AM             I             E         309.00   
2341    01/01/2016 12:00:00 AM             I             E          31.00   
2342    01/01/2016 12:00:00 AM             I             E         311.00   
2343    01/01/2016 12:00:00 AM             I             E         312.00   
2344    01/01/2016 12:00:00 AM             I             E         313.00   
2345    01/01/2016 12:00:00 AM             I             E         319.00   
2346    01/01/2016 12:00:00 AM             I             E          32.00   
2347    01/01/2016 12:00:00 AM             I             E         321.00   
2348    01/01/2016 12:00:00 AM             I             E         326.00   
2349    01/01/2016 12:00:00 AM             I             E          33.00   
2350    01/01/2016 12:00:00 AM             I             E         332.00   
2351    01/01/2016 12:00:00 AM             I             E         339.00   
...                        ...           ...           ...            ...   
624045  01/01/2016 12:00:00 AM             D             F         135.00   
624046  01/01/2016 12:00:00 AM             D             E         160.09   
624047  01/01/2016 12:00:00 AM             D             E         221.02   
624048  01/01/2016 12:00:00 AM             I             E         303.00   
624054  01/01/2016 12:00:00 AM             D             F           1.00   
624056  01/01/2016 12:00:00 AM             D             F         336.00   
624057  01/01/2016 12:00:00 AM             I             E         495.00   
624060  01/01/2016 12:00:00 AM             D             F        1522.00   
624061  01/01/2016 12:00:00 AM             I             E         322.00   
624063  01/01/2016 12:00:00 AM             I             E         612.00   
624064  01/01/2016 12:00:00 AM             D             E          24.00   
624066  01/01/2016 12:00:00 AM             D             F         337.00   
624067  01/01/2016 12:00:00 AM             D             E          85.00   
624069  01/01/2016 12:00:00 AM             D             E         143.00   
624070  01/01/2016 12:00:00 AM             D             E         821.20   
624071  01/01/2016 12:00:00 AM             D             F          33.00   
624072  01/01/2016 12:00:00 AM             D             E         100.00   
624073  01/01/2016 12:00:00 AM             D             E         479.00   
624074  01/01/2016 12:00:00 AM             I             E         421.00   
624075  01/01/2016 12:00:00 AM             D             E         622.00   
624076  01/01/

## Data Investigation

To prepare and investigate the data, there are several things that can be used, for example:

- How many town councils there are? And how much entries has each town council?

- Are there any budgets with value 0?

- What's the relation of of each level (Nivell) and the budget (Import)?

- What are the town councils that had positive/negative balance? (more spent than income)

- Is there any relation between the screen code (Codi pantalla) and the level (Nivell)?

- Do all the town councils use the same amount of levels (Nivell)?

- Do the town councils use the same descriptions (Descripcio) in the different budgets?

- #### How many town councils there are? And how much entries has each town council?

In [76]:
name, counts = np.unique(town_council_data.NOM_COMPLERT, return_counts=True)
print(name.size)
dict(zip(name, counts))

882


{"Ajuntament d'Abrera": 317,
 "Ajuntament d'Agramunt": 832,
 "Ajuntament d'Aguilar de Segarra": 347,
 "Ajuntament d'Agullana": 251,
 "Ajuntament d'Aiguafreda": 275,
 "Ajuntament d'Aiguamúrcia": 830,
 "Ajuntament d'Aiguaviva": 259,
 "Ajuntament d'Aitona": 833,
 "Ajuntament d'Albanyà": 241,
 "Ajuntament d'Albatàrrec": 243,
 "Ajuntament d'Albesa": 832,
 "Ajuntament d'Albinyana": 833,
 "Ajuntament d'Albons": 242,
 "Ajuntament d'Alcanar": 830,
 "Ajuntament d'Alcanó": 833,
 "Ajuntament d'Alcarràs": 834,
 "Ajuntament d'Alcoletge": 833,
 "Ajuntament d'Alcover": 332,
 "Ajuntament d'Aldover": 833,
 "Ajuntament d'Alella": 343,
 "Ajuntament d'Alfara de Carles": 249,
 "Ajuntament d'Alfarràs": 833,
 "Ajuntament d'Alforja": 833,
 "Ajuntament d'Alfés": 832,
 "Ajuntament d'Algerri": 833,
 "Ajuntament d'Alguaire": 833,
 "Ajuntament d'Alió": 255,
 "Ajuntament d'Almacelles": 833,
 "Ajuntament d'Almatret": 833,
 "Ajuntament d'Almenar": 833,
 "Ajuntament d'Almoster": 302,
 "Ajuntament d'Alp": 832,
 "Ajuntam

- #### Are there any budgets with value 0?

In [77]:
town_council_data.loc[town_council_data["IMPORT"] > 0].size

1959716

There are a lot of columns with budget of 0. It may be a good idea to remove them to avoid noise.

- #### What's the relation of of each level (Nivell) and the budget (Import)?

In [78]:
levels = np.unique(town_council_data.NIVELL)
expenses = []
for i in levels:
    current = town_council_data.loc[town_council_data["NIVELL"] == i]
    expenses.append(current["IMPORT"].sum())
expenses

[26356718525.26, 26356718525.26, 25585685635.39, 8272996155.5]

The first and second levels have the same total sum of money, while the third and fourth have different values and each is lower than the previous.

- #### What are the town councils that had positive/negative balance? (more spent than income)

In [90]:
def transform_expense_to_negative(data):
    expenses = data.loc[data["TIPUS_PARTIDA"] == "D"]
    expenses["IMPORT"] = expenses["IMPORT"] * -1
    income = data.loc[data["TIPUS_PARTIDA"] == "I"]
    return expenses.append(income)


transformed_data = town_council_data.copy()
transformed_data = transform_expense_to_negative(transformed_data)

transformed_data = transformed_data.groupby(["NOM_COMPLERT"])["IMPORT"].sum().reset_index()

transformed_data.loc[transformed_data["IMPORT"] < 0]

C:\Users\Jordi\venv\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


NOM_COMPLERT        IMPORT
0                                  Ajuntament d'Abrera -4.999805e+07
1                                Ajuntament d'Agramunt -1.724960e+07
2                      Ajuntament d'Aguilar de Segarra -1.073211e+07
3                                Ajuntament d'Agullana -3.616376e+06
4                              Ajuntament d'Aiguafreda -9.279600e+06
5                             Ajuntament d'Aiguamúrcia -3.205586e+06
6                               Ajuntament d'Aiguaviva -2.501107e+06
7                                  Ajuntament d'Aitona -6.564340e+06
8                                 Ajuntament d'Albanyà -8.801327e+05
9                              Ajuntament d'Albatàrrec -5.522404e+06
10                                 Ajuntament d'Albesa -3.285219e+06
11                              Ajuntament d'Albinyana -7.712288e+06
12                                 Ajuntament d'Albons -2.299100e+06
13                                Ajuntament d'Alcanar -3.402489e+07
14                                 Ajuntament d'Alcanó -1.032279e+06
15                               Ajuntament d'Alcarràs -2.730262e+07
16                              Ajuntament d'Alcoletge -6.887093e+06
17                                Ajuntament d'Alcover -1.900014e+07
18                                Ajuntament d'Aldover -2.003288e+06
19                                 Ajuntament d'Alella -4.660420e+07
20                       Ajuntament d'Alfara de Carles -1.222711e+06
21                               Ajuntament d'Alfarràs -6.557160e+06
22                                Ajuntament d'Alforja -4.110231e+06
23                                  Ajuntament d'Alfés -1.158379e+06
24                                Ajuntament d'Algerri -1.933480e+06
25                               Ajuntament d'Alguaire -7.787262e+06
26                                   Ajuntament d'Alió -1.989433e+06
27                             Ajuntament d'Almacelles -1.604785e+07
28                               Ajuntament d'Almatret -1.223965e+06
29                                Ajuntament d'Almenar -9.003829e+06
..                                                 ...           ...
852                               Ajuntament del Molar -1.758895e+06
853                            Ajuntament del Montmell -7.387060e+06
854                              Ajuntament del Morell -1.878615e+07
855                   Ajuntament del Palau d'Anglesola -6.208745e+06
856                              Ajuntament del Papiol -1.913362e+07
857                             Ajuntament del Perelló -1.077124e+07
858                      Ajuntament del Pinell de Brai -2.639570e+06
859                  Ajuntament del Pla de Santa Maria -6.873800e+06
860                     Ajuntament del Pla del Penedès -3.317859e+06
861                                Ajuntament del Poal -1.715520e+06
862                    Ajuntament del Pont d'Armentera -1.770682e+06
863                         Ajuntament del Pont de Bar -8.951000e+05
864                       Ajuntament del Pont de Suert -8.855805e+06
865         Ajuntament del Pont de Vilomara i Rocafort -1.213140e+07
866                    Ajuntament del Port de la Selva -1.453497e+07
867                   Ajuntament del Prat de Llobregat -3.108411e+08
868                             Ajuntament del Rourell -9.954707e+05
869                             Ajuntament del Soleràs -1.054440e+06
870                            Ajuntament del Vendrell -1.656939e+08
871                            Ajuntament del Vilosell -7.602243e+05
872                             Ajuntament dels Alamús -2.441112e+06
873                          Ajuntament dels Garidells -8.033540e+05
874               Ajuntament dels Hostalets de Pierola -2.087006e+07
875                           Ajuntament dels Omellons -8.327609e+05
876                  Ajuntament dels Omells de na Gaia -8.113010e+05
877                         Ajuntament dels Pallaresos -1.005869e+07
878                       Ajuntament del

Most of the Town Councils (881 over 882) spent more money than they received.

- #### Is there any relation between the screen code (Codi pantalla) and the level (Nivell)?

In [93]:
town_council_data[["TIPUS_PARTIDA", "CODI_PANTALLA", "NIVELL", "NOM_COMPLERT"]]

TIPUS_PARTIDA  CODI_PANTALLA  NIVELL  \
245                D          33.00       2   
671                D          34.00       2   
1501               D         425.00       3   
2325               I           1.00       1   
2326               I          11.00       2   
2327               I         112.00       3   
2328               I         113.00       3   
2329               I         115.00       3   
2330               I         116.00       3   
2331               I          13.00       2   
2332               I         130.00       3   
2333               I           2.00       1   
2334               I          29.00       2   
2335               I         290.00       3   
2336               I           3.00       1   
2337               I          30.00       2   
2338               I         300.00       3   
2339               I         304.00       3   
2340               I         309.00       3   
2341               I          31.00       2   
2342               I         311.00       3   
2343               I         312.00       3   
2344               I         313.00       3   
2345               I         319.00       3   
2346               I          32.00       2   
2347               I         321.00       3   
2348               I         326.00       3   
2349               I          33.00       2   
2350               I         332.00       3   
2351               I         339.00       3   
...              ...            ...     ...   
624045             D         135.00       3   
624046             D         160.09       4   
624047             D         221.02       4   
624048             I         303.00       3   
624054             D           1.00       2   
624056             D         336.00       3   
624057             I         495.00       3   
624060             D        1522.00       4   
624061             I         322.00       3   
624063             I         612.00       3   
624064             D          24.00       2   
624066             D         337.00       3   
624067             D          85.00       2   
624069             D         143.00       3   
624070             D         821.20       4   
624071             D          33.00       2   
624072             D         100.00       4   
624073             D         479.00       3   
624074             I         421.00       3   
624075             D         622.00       3   
624076             D         161.00       3   
624077             D          20.00       2   
624078             D        4312.00       4   
624081             I         339.00       3   
624083             D         124.02       4   
624084             D          34.00       2   
624085             D         226.99       4   
624087             I          11.00       2   
624088             I         392.00       4   
624091             D          44.00       2   

                                  NOM_COMPLERT  
245                      Ajuntament d'Agramunt  
671                     Ajuntament d'Aiguaviva  
1501                     Ajuntament d'Alfarràs  
2325                       Ajuntament d'Abrera  
2326                       Ajuntament d'Abrera  
2327                       Ajuntament d'Abrera  
2328                       Ajuntament d'Abrera  
2329                       Ajuntament d'Abrera  
2330                       Ajuntament d'Abrera  
2331                       Ajuntament d'Abrera  
2332                       Ajuntament d'Abrera  
2333                       Ajuntament d'Abrera  
2334                       Ajuntament d'Abrera  
2335                       Ajuntament d'Abrera  
2336                       Ajuntament d'Abrera  
2337                       Ajuntament d'Abrera  
2338                       Ajuntament d'Abrera  
2339                       Ajuntament d'Abrera  
2340                       Ajuntament d'Abrera  
2341                       Ajuntament d'Abrera  
2342                       Ajuntament d'Abrera  
2343          

As we can see in the table, when the level (Nivell) is 1, the screen code (Codi_pantalla) has only one digit. The same happens when the level is 2 or 3. But when the level is 4, the screen code can have either 4 digits or use the decimals. 

- #### Do all the town councils use the same amount of levels (Nivell)?

In [139]:
town_council_names = np.unique(town_council_data.NOM_COMPLERT)

level_codes = []
for name in town_council_names:
    num_codes = np.unique(town_council_data.NIVELL).shape[0]
    level_codes.append([name, num_codes])

level_codes

[["Ajuntament d'Abrera", 4],
 ["Ajuntament d'Agramunt", 4],
 ["Ajuntament d'Aguilar de Segarra", 4],
 ["Ajuntament d'Agullana", 4],
 ["Ajuntament d'Aiguafreda", 4],
 ["Ajuntament d'Aiguamúrcia", 4],
 ["Ajuntament d'Aiguaviva", 4],
 ["Ajuntament d'Aitona", 4],
 ["Ajuntament d'Albanyà", 4],
 ["Ajuntament d'Albatàrrec", 4],
 ["Ajuntament d'Albesa", 4],
 ["Ajuntament d'Albinyana", 4],
 ["Ajuntament d'Albons", 4],
 ["Ajuntament d'Alcanar", 4],
 ["Ajuntament d'Alcanó", 4],
 ["Ajuntament d'Alcarràs", 4],
 ["Ajuntament d'Alcoletge", 4],
 ["Ajuntament d'Alcover", 4],
 ["Ajuntament d'Aldover", 4],
 ["Ajuntament d'Alella", 4],
 ["Ajuntament d'Alfara de Carles", 4],
 ["Ajuntament d'Alfarràs", 4],
 ["Ajuntament d'Alforja", 4],
 ["Ajuntament d'Alfés", 4],
 ["Ajuntament d'Algerri", 4],
 ["Ajuntament d'Alguaire", 4],
 ["Ajuntament d'Alió", 4],
 ["Ajuntament d'Almacelles", 4],
 ["Ajuntament d'Almatret", 4],
 ["Ajuntament d'Almenar", 4],
 ["Ajuntament d'Almoster", 4],
 ["Ajuntament d'Alp", 4],
 ["Ajunta

All the town councils use the same number of levels.

- #### Do the town councils use the same descriptions (Descripcio) in the different budgets?

In [153]:
town_council_codes = np.unique(town_council_data.CODI_ENS)

description_num = np.empty((town_council_codes.shape[0], 2))
i = 0
for code in town_council_codes:
    descriptions = town_council_data.loc[town_council_data["CODI_ENS"] == code]
    num_description = np.unique(descriptions.DESCRIPCIO).shape[0]
    description_num[i] = [code, num_description]
    i += 1

pd.DataFrame(description_num, columns=["CODE", "DESCRIPTION_NUM"]).sort_values(["DESCRIPTION_NUM"])

CODE  DESCRIPTION_NUM
170  8.187830e+08            168.0
847  4.314950e+09            198.0
46   8.057500e+08            200.0
69   8.080400e+08            203.0
688  2.524770e+09            205.0
479  1.722290e+09            206.0
418  1.715760e+09            206.0
528  2.504500e+09            207.0
83   8.095800e+08            208.0
437  1.717660e+09            209.0
413  1.715180e+09            213.0
482  1.722530e+09            214.0
385  1.711940e+09            215.0
81   8.093000e+08            215.0
36   8.045900e+08            217.0
633  2.517520e+09            217.0
137  8.150900e+08            217.0
783  4.308030e+09            218.0
381  1.711560e+09            219.0
429  1.716800e+09            219.0
823  4.312270e+09            219.0
451  1.719250e+09            219.0
487  1.723270e+09            220.0
118  8.130400e+08            220.0
204  8.225100e+08            221.0
160  8.177000e+08            221.0
297  1.701600e+09            221.0
308  1.702800e+09            222.0
475  1.721740e+09            222.0
441  1.718270e+09            222.0
..            ...              ...
655  2.520560e+09            648.0
654  2.520430e+09            648.0
549  2.507070e+09            648.0
652  2.520220e+09            648.0
658  2.520810e+09            648.0
668  2.522060e+09            648.0
452  1.719310e+09            648.0
678  2.523110e+09            648.0
684  2.524240e+09            648.0
683  2.524020e+09            648.0
682  2.523980e+09            648.0
543  2.506210e+09            648.0
680  2.523470e+09            648.0
669  2.522130e+09            648.0
677  2.523040e+09            648.0
453  1.719460e+09            648.0
544  2.506370e+09            648.0
675  2.522710e+09            648.0
674  2.522650e+09            648.0
673  2.522520e+09            648.0
545  2.506420e+09            648.0
670  2.522280e+09            648.0
296  1.701570e+09            649.0
90   8.102200e+08            649.0
706  2.591180e+09            649.0
501  2.501170e+09            649.0
609  2.514840e+09            649.0
535  2.505390e+09            649.0
851  4.315360e+09            649.0
163  8.180300e+08            649.0

[882 rows x 2 columns]

We can see that the town councils have a very different number of descriptions, going from 168 to 649.

## Clustering ideas to explore or try

- Use the rows with the same level (Nivell) regardless of any other value. Does the result make any sense?
- Cluster the town councils using a new column that days if the balance is positive or negative.
- Will we get the same amount of clusters than different levels (Nivell) if we use the column screen code (Codi pantalla) for the clustering?
- Some of the previous clusterings can be used for geolocalization? Ex. giving the zone where the cluster is found.